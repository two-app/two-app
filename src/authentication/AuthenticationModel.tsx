export type Tokens = {
  accessToken: string;
  refreshToken: string;
};

export type UserRegistration = {
  uid: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  dob: string;
  acceptedTerms: boolean;
};
