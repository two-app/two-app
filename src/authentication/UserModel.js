// @flow

import jwt_decode from "jwt-decode";

const UnconnectedUser: {
    uid: number,
    connectCode: string
} = {
    uid: null,
    connectCode: null
};

const createUnconnectedUser = (uid: number, connectCode: string): UnconnectedUser => ({uid, connectCode});

const parseUnconnectedUserFromToken = (accessToken: string): (UnconnectedUser) => {
    const decoded = jwt_decode(accessToken);
    const uid = decoded['uid'];
    const connectCode = decoded['connectCode'];

    return createUnconnectedUser(uid, connectCode);
};

export {UnconnectedUser, createUnconnectedUser, parseUnconnectedUserFromToken};