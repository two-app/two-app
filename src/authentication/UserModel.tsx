import jwt_decode from 'jwt-decode';

// @ts-ignore
const decode = (token: string): any => jwt_decode(token);

export type UnconnectedUser = {
    uid: number,
    connectCode: string
}

export const unconnectedUserFromAccessToken = (accessToken: string): UnconnectedUser => ({
    uid: decode(accessToken)['uid'],
    connectCode: decode(accessToken)['connectCode']
});

export type User = {
    uid: number,
    pid: number,
    cid: number
}

export const userFromAccessToken = (accessToken: string): User => ({
    uid: decode(accessToken)['uid'],
    pid: decode(accessToken)['pid'],
    cid: decode(accessToken)['cid']
});

export const detectUserFromAccessToken = (accessToken: string): (User | UnconnectedUser) => {
    const role: string = decode(accessToken)['role'];
    if (role.toLowerCase() == 'connect') {
        return unconnectedUserFromAccessToken(accessToken);
    } else {
        return userFromAccessToken(accessToken);
    }
}

/**
 * @param user {User | UnconnectedUser}  narrows the type based if the connect code is present.
 */
export const isUnconnectedUser = (user: User | UnconnectedUser): user is UnconnectedUser => {
    return (user as UnconnectedUser).connectCode !== undefined;
}