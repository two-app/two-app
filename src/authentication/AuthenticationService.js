// @flow
import Gateway from "../http/Gateway";
import {AxiosError, AxiosResponse} from "axios";
import {UserRegistration} from "./register_workflow/UserRegistrationModel";
import {UnconnectedUser} from "./UserModel";
import {Tokens} from "./AuthenticationModel";

class RegisterUserResponse {
    user: UnconnectedUser;
    tokens: Tokens;

    constructor(user: UnconnectedUser, tokens: Tokens) {
        this.user = user;
        this.tokens = tokens;
    }
}


const registerUser = (userRegistration: UserRegistration): Promise<RegisterUserResponse> =>
    Gateway.post("/self", userRegistration)
        .then((r: AxiosResponse) => {
            const accessToken = r.data['accessToken'];
            const refreshToken = r.data['refreshToken'];
            return new RegisterUserResponse({
                user: UnconnectedUser.fromAccessToken(accessToken),
                tokens: new Tokens(accessToken, refreshToken)
            });
        }).catch((e: AxiosError) => {
            throw new Error(e.response.data['message'].toString());
        }
    );

export default {registerUser};
export {RegisterUserResponse};