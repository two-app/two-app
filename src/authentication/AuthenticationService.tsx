import Gateway from '../http/Gateway';
import {AxiosError, AxiosResponse} from 'axios';
import UserRegistrationModel, {UserRegistration} from './register_workflow/UserRegistrationModel';
import {UnconnectedUser, unconnectedUserFromAccessToken, User, userFromAccessToken, detectUserFromAccessToken} from './UserModel';
import {Tokens} from './AuthenticationModel';

export type UserResponse = {
    user: UnconnectedUser | User,
    tokens: Tokens
}

export type LoginCredentials = {
    email: string,
    rawPassword: string
}

export const areCredentialsValid = ({ email, rawPassword }: LoginCredentials) =>
    UserRegistrationModel.isEmailValid(email) && rawPassword.length > 3;

const login = (loginCredentials: LoginCredentials): Promise<UserResponse> => Gateway.post("/login", loginCredentials)
    .then((r: AxiosResponse<Tokens>): UserResponse => ({
        user: detectUserFromAccessToken(r.data.accessToken),
        tokens: {accessToken: r.data.accessToken, refreshToken: r.data.refreshToken}
    })).catch((e: AxiosError) => {
        throw new Error(e.response?.data['reason'].toString());
    });

const registerUser = (userRegistration: UserRegistration): Promise<UserResponse> => Gateway.post('/self', userRegistration)
    .then((r: AxiosResponse<Tokens>): UserResponse => ({
        user: unconnectedUserFromAccessToken(r.data.accessToken),
        tokens: {accessToken: r.data.accessToken, refreshToken: r.data.refreshToken}
    })).catch((e: AxiosError) => {
        throw new Error(e.response?.data['reason'].toString());
    });

const connectToPartner = (connectCode: String): Promise<UserResponse> => Gateway.post(`/partner/${connectCode}`)
    .then((r: AxiosResponse<Tokens>): UserResponse => ({
        user: userFromAccessToken(r.data.accessToken),
        tokens: {accessToken: r.data.accessToken, refreshToken: r.data.refreshToken}
    })).catch((e: AxiosError) => {
        // @ts-ignore
        throw new Error(e.response.data['reason'].toString());
    });

export default {login, registerUser, connectToPartner};