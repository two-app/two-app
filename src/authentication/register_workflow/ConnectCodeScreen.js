// @flow

import React, {useState} from "react";
import {ActivityIndicator, Clipboard, Dimensions, StyleSheet, Text, TouchableOpacity, View} from "react-native";
import {connect} from "react-redux";
import {WrapperContainer} from "../../views/View";
import LogoHeader from "../LogoHeader";
import {UnconnectedUser} from "../UserModel";
import Colors from "../../Colors";
import Input from "../../forms/Input";
import SubmitButton from "../../forms/SubmitButton";
import LoadingView from "../../views/LoadingView";
import Gateway from "../../http/Gateway";

/**
 * @param navigation
 * @param user {UnconnectedUser}
 */
const ConnectCodeScreen = ({user}) => {
    const [partnerConnectCode, setPartnerConnectCode] = useState("");
    const isPartnerCodeValid = partnerCode => partnerCode.length === 6 && partnerCode !== user.connectCode;
    const [submitted, setSubmitted] = useState(false);

    if (submitted) {
        Gateway.get("/partner").then(console.log).catch(console.error);
    }

    return <>
        {submitted && <LoadingView/>}
        <WrapperContainer>
            <LogoHeader heading="Connect Your Partner"/>
            <Text style={styles.subheading}>Thanks for joining us!</Text>
            <Text style={styles.paragraph}>
                The sign-up process is almost complete. Once your partner has registered, send them your code!
            </Text>
            <TouchableOpacity style={styles.copyButton} onPress={() => Clipboard.setString(user.connectCode)}>
                <Text style={{...styles.copyTip, marginBottom: 10}}>Your Code</Text>
                <Text style={styles.code}>{user.connectCode}</Text>
                <Text style={{...styles.copyTip, marginTop: 10}}>Tap to Copy</Text>
            </TouchableOpacity>

            <Text style={{...styles.subheading, marginTop: 20}}>Or, enter your partners code...</Text>
            <View style={styles.codeInputContainer}>
                <Input attributes={{placeholder: "e.g bWzGl2"}}
                       isValid={() => isPartnerCodeValid(partnerConnectCode)}
                       onChange={setPartnerConnectCode}
                />
                {partnerConnectCode === user.connectCode &&
                <Text style={styles.error} id="error">You can't connect with yourself!</Text>
                }
                <SubmitButton onSubmit={() => setSubmitted(true)} text="Connect"
                              disabled={!isPartnerCodeValid(partnerConnectCode)}/>
            </View>
        </WrapperContainer>
    </>;
};

ConnectCodeScreen.navigationOptions = {
    title: 'Partner Connect',
    header: null
};

const styles = StyleSheet.create({
    subheading: {
        marginTop: 20,
        fontWeight: "bold"
    },
    paragraph: {
        marginTop: 10
    },
    copyButton: {
        marginTop: 20,
        backgroundColor: Colors.LIGHT,
        padding: 20
    },
    copyTip: {
        fontWeight: "300",
        color: Colors.DARK,
        textAlign: "center"
    },
    code: {
        fontWeight: "bold",
        fontSize: 30,
        textAlign: "center"
    },
    codeInputContainer: {
        marginTop: 5
    },
    codeInput: {
        padding: 20,
        borderBottomWidth: 1,
        borderBottomColor: Colors.DARK
    },
    error: {
        color: Colors.DARK_SALMON,
        marginTop: 10
    },
    overlay: {
        flex: 1,
        position: "absolute",
        opacity: 0.5,
        backgroundColor: 'white',
        width: Dimensions.get("window").width,
        height: Dimensions.get("window").height,
        zIndex: 2,
        alignItems: "center",
        justifyContent: "center"
    },
});

const mapStateToProps = state => ({user: state['user']});

export default connect(mapStateToProps)(ConnectCodeScreen);
export {ConnectCodeScreen};