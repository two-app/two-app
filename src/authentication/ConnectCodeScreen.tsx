import {validate as isUuid} from 'uuid';
import {v4 as uuid} from 'uuid';
import {useState} from 'react';
import {
  RefreshControl,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import Clipboard from '@react-native-community/clipboard';
import HapticFeedback from 'react-native-haptic-feedback';
import {ScrollContainer} from '../views/View';
import {LogoHeader} from './LogoHeader';
import Colors from '../Colors';
import {Input} from '../forms/Input';
import {resetNavigate, Screen} from '../navigation/NavigationUtilities';
import ConnectService from '../user/ConnectService';
import type {ErrorResponse} from '../http/Response';
import AuthenticationService from './AuthenticationService';
import type {User} from './UserModel';
import {PrimaryButton} from '../forms/SubmitButton';
import IonIcon from 'react-native-vector-icons/Ionicons';
import {HR} from '../forms/HorizontalRule';
import F, {Form} from '../forms/Form';
import moment from 'moment';
import {useAuthStore} from './AuthenticationStore';

type ConnectForm = {
  toUser: string;
  anniversary: string;
};

export const ConnectCodeScreen = ({
  navigation,
}: Screen<'ConnectCodeScreen'>) => {
  const [form, setForm] = useState<Form<ConnectForm>>({
    toUser: F.str,
    anniversary: F.str,
  });
  const [submitted, setSubmitted] = useState(false);
  const [refreshing, setRefreshing] = useState(false);
  const [error, setError] = useState<string>();

  const authStore = useAuthStore();

  const connectToPartner = () => {
    setSubmitted(true);
    const data = F.data(form);
    AuthenticationService.connectUser({...data, cid: uuid()})
      .then((_: User) => resetNavigate('HomeScreen', navigation))
      .catch(({reason}: ErrorResponse) => {
        if (
          reason === 'User already has a partner.' ||
          reason === 'Invalid JSON Web Token'
        ) {
          refresh();
        } else {
          setError(reason);
          setSubmitted(false);
        }
      });
  };

  const refresh = () => {
    ConnectService.checkConnection()
      .then((isConnected: boolean) => {
        if (isConnected) {
          resetNavigate('HomeScreen', navigation);
        } else {
          setRefreshing(false);
        }
      })
      .catch((e: ErrorResponse) => {
        setError(e.reason);
        setSubmitted(false);
        setRefreshing(false);
      });
  };

  return (
    <ScrollContainer
      refreshControl={
        <RefreshControl
          colors={['#9Bd35A', '#689F38']}
          refreshing={refreshing}
          onRefresh={() => {
            setRefreshing(true);
            refresh();
          }}
        />
      }>
      <LogoHeader heading="Connect To Your Partner" />
      <Text style={{marginTop: 10}}>
        Send your code, or enter theirs below.
      </Text>

      <CopyConnectCodeButton code={authStore.user!.uid} />

      <HR />

      <Text style={{...styles.subheading}}>
        Partner's Code & Anniversary Date
      </Text>

      <Input
        placeholder="Partner's Connect Code"
        accessibilityLabel="Partner Code"
        onEmit={toUser => setForm({...form, toUser})}
        isValid={isUuid}
        autoCorrect={false}
        icon={{provider: IonIcon, name: 'finger-print-outline'}}
        containerStyle={{marginTop: 20}}
      />

      <Text style={styles.inputHint}>YYYY-MM-DD</Text>
      <Input
        placeholder="Anniversary Date"
        accessibilityLabel="Anniversary Date"
        onEmit={anniversary => setForm({...form, anniversary})}
        isValid={dob => {
          const m = moment(dob, 'YYYY-MM-DD', true);
          return m.isValid() && m.year() <= new Date().getFullYear();
        }}
        returnKeyType="done"
        keyboardType="numeric"
        mask={{
          type: 'datetime',
          options: {format: 'YYYY-MM-DD'},
        }}
        icon={{provider: IonIcon, name: 'calendar-outline'}}
      />

      {error != '' && <Text style={styles.error}>{error}</Text>}

      <PrimaryButton
        accessibilityLabel="Connect to Partner"
        loading={submitted}
        onPress={connectToPartner}
        disabled={F.isInvalid(form)}>
        Connect to Partner <IonIcon name="heart" />
      </PrimaryButton>

      <TouchableOpacity
        accessibilityLabel="Logout"
        onPress={() => navigation.navigate('LogoutScreen')}
        style={{marginTop: 40}}>
        <Text style={{color: Colors.REGULAR, fontWeight: '500'}}>Logout</Text>
      </TouchableOpacity>
    </ScrollContainer>
  );
};

const CopyConnectCodeButton = ({code}: {code: string}) => {
  const [copied, setCopied] = useState(false);

  const onCopy = () => {
    Clipboard.setString(code);
    HapticFeedback.trigger('selection', {enableVibrateFallback: false});
    setCopied(true);
  };

  const buttonColor = copied ? Colors.VALID_GREEN : Colors.LIGHT;
  const textColor = copied ? 'white' : Colors.REGULAR;
  const codeColor = copied ? 'white' : Colors.DARK;

  return (
    <TouchableOpacity
      onPress={onCopy}
      style={[styles.copyButton, {backgroundColor: buttonColor}]}
      accessibilityHint="Copy connect code to clipboard"
      accessibilityState={{checked: copied}}>
      <Text style={[styles.copyTip, {color: textColor, marginBottom: 10}]}>
        {copied ? 'Copied!' : 'Tap to Copy'}
      </Text>

      <Text style={[styles.code, {color: codeColor}]}>{code}</Text>

      <View style={{marginTop: 10}}>
        <IonIcon
          name="finger-print-outline"
          style={[styles.copyTip, {color: textColor}]}
          size={20}
        />
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  subheading: {
    fontWeight: '600',
    color: Colors.DARK,
    textAlign: 'center',
  },
  copyButton: {
    marginTop: 20,
    padding: 20,
  },
  inputHint: {
    color: Colors.REGULAR,
    fontSize: 10,
    paddingTop: 15,
    paddingBottom: 5,
    paddingLeft: 1,
  },
  copyTip: {
    fontWeight: '500',
    textAlign: 'center',
  },
  code: {
    fontWeight: 'bold',
    fontSize: 25,
    textAlign: 'center',
  },
  error: {
    color: Colors.DARK_SALMON,
    marginTop: 10,
  },
});
