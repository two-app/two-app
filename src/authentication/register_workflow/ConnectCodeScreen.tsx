import React, {useState} from 'react';
import {Clipboard, StyleSheet, Text, TouchableOpacity, View} from 'react-native';
import {connect, ConnectedProps} from 'react-redux';
import {WrapperContainer} from '../../views/View';
import LogoHeader from '../LogoHeader';
import {User} from '../UserModel';
import Colors from '../../Colors';
import Input from '../../forms/Input';
import SubmitButton from '../../forms/SubmitButton';
import LoadingView from '../../views/LoadingView';
import AuthenticationService, {UserResponse} from '../AuthenticationService';
import {NavigationActions, StackActions} from 'react-navigation';
import {NavigationStackProp} from 'react-navigation-stack';
import {TwoState} from '../../state/reducers';
import {selectUnconnectedUser, storeUser} from '../../user';
import {storeTokens} from '../store';


const mapState = (state: TwoState) => ({user: selectUnconnectedUser(state.user)});
const mapDispatch = {storeUser, storeTokens};
const connector = connect(mapState, mapDispatch);
type ConnectorProps = ConnectedProps<typeof connector>;
type ConnectCodeScreenProps = ConnectorProps & {
    navigation: NavigationStackProp
};

const ConnectCodeScreen = ({navigation, user, storeUser, storeTokens}: ConnectCodeScreenProps) => {
    const [partnerConnectCode, setPartnerConnectCode] = useState('');
    const isPartnerCodeValid = (partnerCode: string) => partnerCode.length === 6 && partnerCode !== user.connectCode;
    const [submitted, setSubmitted] = useState(false);
    const [error, setError] = useState<string | null>(null);

    const navigateToHomeScreen = () => navigation.dispatch(
        StackActions.reset({
            index: 0,
            actions: [NavigationActions.navigate({routeName: 'HomeScreen'})]
        })
    );

    const connectToPartner = (connectCode: string) => {
        setSubmitted(true);
        AuthenticationService.connectToPartner(connectCode).then((response: UserResponse) => {
            storeUser(response.user as User);
            storeTokens(response.tokens);
            navigateToHomeScreen();
        }).catch((e: Error) => {
            setSubmitted(false);
            setError(e.message);
        });
    };

    return <>
        {submitted && <LoadingView/>}
        <WrapperContainer>
            <LogoHeader heading="Connect Your Partner"/>
            <Text style={styles.subheading}>Thanks for joining us!</Text>
            <Text style={styles.paragraph}>
                The sign-up process is almost complete. Once your partner has registered, send them your code!
            </Text>
            <TouchableOpacity style={styles.copyButton} onPress={() => Clipboard.setString(user.connectCode)}>
                <Text style={{...styles.copyTip, marginBottom: 10}}>Your Code</Text>
                <Text style={styles.code}>{user.connectCode}</Text>
                <Text style={{...styles.copyTip, marginTop: 10}}>Tap to Copy</Text>
            </TouchableOpacity>

            <Text style={{...styles.subheading, marginTop: 20}}>Or, enter your partners code...</Text>
            <View style={styles.codeInputContainer}>
                <Input attributes={{placeholder: 'e.g bWzGl2'}}
                       isValid={() => isPartnerCodeValid(partnerConnectCode)}
                       onChange={setPartnerConnectCode}
                />
                {partnerConnectCode === user.connectCode &&
                <Text style={styles.error} data-testid="error">You can't connect with yourself!</Text>
                }
                {error && <Text style={styles.error} data-testid="error">{error}</Text>}
                <SubmitButton onSubmit={() => connectToPartner(partnerConnectCode)} text="Connect"
                              disabled={!isPartnerCodeValid(partnerConnectCode)}/>
            </View>
        </WrapperContainer>
    </>;
};

ConnectCodeScreen.navigationOptions = {
    title: 'Partner Connect',
    header: null
};

const styles = StyleSheet.create({
    subheading: {
        marginTop: 20,
        fontWeight: 'bold'
    },
    paragraph: {
        marginTop: 10
    },
    copyButton: {
        marginTop: 20,
        backgroundColor: Colors.LIGHT,
        padding: 20
    },
    copyTip: {
        fontWeight: '300',
        color: Colors.DARK,
        textAlign: 'center'
    },
    code: {
        fontWeight: 'bold',
        fontSize: 30,
        textAlign: 'center'
    },
    codeInputContainer: {
        marginTop: 5
    },
    codeInput: {
        padding: 20,
        borderBottomWidth: 1,
        borderBottomColor: Colors.DARK
    },
    error: {
        color: Colors.DARK_SALMON,
        marginTop: 10
    }
});

export default connector(ConnectCodeScreen);
export {ConnectCodeScreen};