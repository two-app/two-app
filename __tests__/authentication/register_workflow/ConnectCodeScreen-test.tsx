import 'react-native';
import { Clipboard } from 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { UnconnectedUser } from '../../../src/authentication/UserModel';
import { ConnectCodeScreen } from '../../../src/authentication/register_workflow/ConnectCodeScreen';
import AuthenticationService, { UserResponse } from '../../../src/authentication/AuthenticationService';
import { CommonActions } from '@react-navigation/native';

describe('ConnectCodeScreen', () => {
    let tb: ConnectCodeScreenTestBed;
    beforeEach(() => tb = new ConnectCodeScreenTestBed());

    test('should maintain snapshot', () => expect(renderer.create(
        <ConnectCodeScreen navigation={{} as any}
            storeTokens={jest.fn()}
            storeUser={jest.fn()}
            user={tb.user} />
    )).toMatchSnapshot());

    describe('Tapping the Connect Code', () => {
        beforeEach(() => tb.clickCopyToClipboard());

        test('puts code in clipboard', () => expect(tb.clipboardSetStringFn).toHaveBeenCalledWith(tb.user.connectCode));
    });

    describe('Clicking the logout button', () => {
        test('it should navigate to the LogoutScreen', () => {
            tb.clickLogout();

            expect(tb.dispatchFn).toHaveBeenCalledTimes(1);
        });
    });

    describe('By Default', () => {
        test('submit is disabled', () => expect(tb.isSubmitButtonDisabled()).toBe(true));
    });

    describe('When entered partner code is identical to users code', () => {
        beforeEach(() => tb.setPartnerCodeInput(tb.user.connectCode));

        test('displays error', () => expect(
            tb.getErrorText()
        ).toEqual('You can\'t connect with yourself!'));

        test('validates input', () => expect(tb.isPartnerCodeValid()).toBe(false));

        test('disables submit', () => expect(tb.isSubmitButtonDisabled()).toBe(true));
    });

    describe('When entered code is not 6 characters in length', () => {
        beforeEach(() => tb.setPartnerCodeInput('ab'));

        test('disables submit', () => expect(tb.isSubmitButtonDisabled()).toBe(true));
    });

    describe('When entered code is valid', () => {
        beforeEach(() => tb.setPartnerCodeInput('ghijkl'));

        test('enables submit', () => expect(tb.isSubmitButtonDisabled()).toBe(false));

        test('displays loading view on submit', () => {
            tb.clickSubmit();
            expect(tb.wrapper.find("ScrollContainer").prop<boolean>("isLoading")).toBe(true);
        });

        describe('On successful connect', () => {
            const response: UserResponse = {
                user: { uid: 1, pid: 2, cid: 3 },
                tokens: { accessToken: 'testAccess', refreshToken: 'testRefresh' }
            };

            beforeEach(() => {
                tb.whenConnectResolve(response);
                tb.clickSubmit();
            });

            test('stores user in redux', done => setImmediate(() => {
                expect(tb.storeUserFn).toHaveBeenCalledWith({ ...response.user });
                done();
            }));

            test('stores tokens in redux', done => setImmediate(() => {
                expect(tb.setTokensFn).toHaveBeenCalledWith({ ...response.tokens });
                done();
            }));

            test('navigates to Home Screen', done => setImmediate(() => {
                expect(tb.dispatchFn).toHaveBeenCalledWith(
                    CommonActions.reset({
                        index: 0,
                        routes: [{ name: 'HomeScreen' }]
                    })
                );
                done();
            }));
        });

        describe('On failed connect', () => {
            const response = new Error('Error Message');
            beforeEach(() => {
                tb.whenConnectReject(response);
                tb.clickSubmit();
            });

            test('hides loading view', () => expect(tb.wrapper.exists('LoadingView')).toBe(false));

            test('displays error message', () => expect(tb.getErrorText()).toEqual(response.message));
        });
    });
});

class ConnectCodeScreenTestBed {
    user: UnconnectedUser = { uid: 12, connectCode: 'abcdef' };
    clipboardSetStringFn = jest.fn();
    setTokensFn = jest.fn();
    storeUserFn = jest.fn();
    dispatchFn = jest.fn();

    wrapper = shallow(
        <ConnectCodeScreen user={this.user}
            storeTokens={this.setTokensFn}
            storeUser={this.storeUserFn}
            navigation={{ dispatch: this.dispatchFn } as any}
        />
    );

    constructor() {
        Clipboard.setString = this.clipboardSetStringFn;
    }

    clickCopyToClipboard = () => this.wrapper.find('TouchableOpacity').prop<() => void>('onPress')();
    setPartnerCodeInput = (v: string) => this.wrapper.find('Input').prop<(v: string) => void>('onChange')(v);
    isPartnerCodeValid = () => this.wrapper.find('Input').prop<() => boolean>('isValid')();
    isSubmitButtonDisabled = () => this.wrapper.find('SubmitButton').prop<boolean>('disabled');
    clickSubmit = () => this.wrapper.find('SubmitButton').prop<() => void>('onSubmit')();
    getErrorText = () => this.wrapper.find('Text[data-testid=\'error\']').render().text();
    clickLogout = () => this.wrapper.find('Button[data-testid=\'logout-button\']').prop<() => void>("onClick")();

    whenConnectResolve = (userResponse: UserResponse) => {
        AuthenticationService.connectToPartner = jest.fn().mockResolvedValue(userResponse);
    };

    whenConnectReject = (error: Error) => {
        AuthenticationService.connectToPartner = jest.fn().mockRejectedValue(error);
    };
}