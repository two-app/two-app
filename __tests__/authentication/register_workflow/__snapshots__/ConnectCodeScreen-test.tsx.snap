// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectCodeScreen should maintain snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "flex": 1,
            "marginLeft": "5%",
            "marginRight": "5%",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#EE7674",
              "fontSize": 50,
              "fontWeight": "bold",
              "paddingBottom": 20,
              "paddingTop": 20,
              "textAlign": "center",
            }
          }
        >
          two.
        </Text>
        <Text
          data-testid="heading"
          style={
            Object {
              "color": "rgb(37, 37, 37)",
              "fontSize": 25,
              "fontWeight": "bold",
            }
          }
        >
          Connect Your Partner
        </Text>
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Thanks for joining us!
        </Text>
        <Text
          style={
            Object {
              "marginTop": 10,
            }
          }
        >
          The sign-up process is almost complete. Once your partner has registered, send them your code!
        </Text>
        <TouchableOpacity
          onPress={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(0, 0, 0, 0.1)",
              "marginTop": 20,
              "padding": 20,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "rgb(37, 37, 37)",
                "fontWeight": "300",
                "marginBottom": 10,
                "textAlign": "center",
              }
            }
          >
            Your Code
          </Text>
          <Text
            style={
              Object {
                "fontSize": 30,
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            abcdef
          </Text>
          <Text
            style={
              Object {
                "color": "rgb(37, 37, 37)",
                "fontWeight": "300",
                "marginTop": 10,
                "textAlign": "center",
              }
            }
          >
            Tap to Copy
          </Text>
        </TouchableOpacity>
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Or, enter your partners code...
        </Text>
        <View
          style={
            Object {
              "marginTop": 5,
            }
          }
        >
          <TextInput
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="e.g bWzGl2"
            style={
              Array [
                Object {
                  "borderBottomColor": "rgba(0, 0, 0, 0.2)",
                  "borderBottomWidth": 1,
                  "fontSize": 15,
                  "paddingBottom": 10,
                  "paddingHorizontal": 0,
                  "paddingTop": 10,
                  "width": "100%",
                },
                undefined,
                undefined,
              ]
            }
            value=""
          />
          <TouchableOpacity
            disabled={true}
            onPress={[Function]}
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#EE7674",
                  "justifyContent": "center",
                  "marginBottom": 25,
                  "marginTop": 25,
                  "padding": 20,
                },
                Object {
                  "backgroundColor": "rgba(0, 0, 0, 0.2)",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontWeight": "bold",
                }
              }
            >
              Connect
            </Text>
          </TouchableOpacity>
        </View>
        <TouchableOpacity
          onPress={[Function]}
        >
          <Text>
            LOGOUT
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
